driver-core/src/main/com/mongodb/internal/session/ServerSessionPool.java

```diff
@@ -118,7 +122,23 @@ private void endClosedSessions() {
             return;
         }
 
-        Connection connection = cluster.selectServer(new ReadPreferenceServerSelector(ReadPreference.primaryPreferred())).getConnection();
+        final List<ServerDescription> primaryPreferred = new ReadPreferenceServerSelector(ReadPreference.primaryPreferred())
```

Comment:
Use the server selector directly on the current cluster description so that we avoid blocking in selectServer in the case where there is no server matching this read preference.  Then if any servers are available select the first one regardless of its ok status, so that it doesn't block under most circumstances.

The only way this blocks now is if between the call to get the current cluster description and the call to select server, the server that we choose entirely disappears from the list (e.g. it's taken out of the replica set configuration).